beforeSpecRun is not a function

Cypress 3.1.0 Linux CircleCI job https://circleci.com/gh/cypress-io/cypress/33960

```
Warning: It looks like you are trying to record this run from a forked PR.

The 'Record Key' is missing. Your CI provider is likely not passing private environment variables to builds from forks.

These results will not be recorded.

This error will not alter the exit code.

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    3.1.0                                                                              │
  │ Browser:    Electron 59 (headless)                                                             │
  │ Specs:      16 found (app_spec.coffee, fixtures_spec.coffee, footer_spec.coffee, global_mode_… │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘

beforeSpecRun is not a function
TypeError: beforeSpecRun is not a function
    at parallelWithRecord (/root/cypress/packages/server/lib/modes/run.coffee:282:5)
    at iterateThroughSpecs (/root/cypress/packages/server/lib/modes/run.coffee:307:7)
    at Object.runSpecs (/root/cypress/packages/server/lib/modes/run.coffee:789:5)
    at runAllSpecs (/root/cypress/packages/server/lib/modes/run.coffee:953:12)
    at /root/cypress/packages/server/lib/modes/record.coffee:451:9
From previous event:
    at /root/cypress/packages/server/lib/modes/record.coffee:449:6
    at runCallback (timers.js:781:20)
    at tryOnImmediate (timers.js:743:5)
    at processImmediate [as _immediateCallback] (timers.js:714:5)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @packages/desktop-gui@0.0.0 cypress:run: `node ../../scripts/cypress run --project . "--record" "--parallel" "--group" "2x-desktop-gui"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @packages/desktop-gui@0.0.0 cypress:run script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -186,7 +186,11 @@ This outputs a lot of debugging lines. To focus on an individual module, run wit
 
 When running `npm start` this routes through the CLI and eventually calls `npm run dev` with the proper arguments. This enables Cypress day-to-day development to match the logic of the built binary + CLI integration.
 
-If you want to bypass the CLI entirely, you can use the `npm run dev` task and pass arguments directly.
+If you want to bypass the CLI entirely, you can use the `npm run dev` task and pass arguments directly. For example to headlessly run a project in a given folder, while trying to record to the Dashboard
+
+```text
+npm run dev -- --run-project /project/folder --record --key <key>
+```
 
 #### Tasks
 

--- a/packages/server/lib/modes/record.coffee
+++ b/packages/server/lib/modes/record.coffee
@@ -448,7 +448,9 @@ createRunAndRecordSpecs = (options = {}) ->
     })
     .then (resp) ->
       if not resp
-        runAllSpecs()
+        ## if a forked run, can't record and can't be parallel
+        ## because the necessary env variables aren't present
+        runAllSpecs({}, false)
       else
         { runUrl, runId, machineId, groupId } = resp
 
@@ -529,7 +531,7 @@ createRunAndRecordSpecs = (options = {}) ->
                 instanceId
               })
 
-        runAllSpecs(beforeSpecRun, afterSpecRun, runUrl)
+        runAllSpecs({ beforeSpecRun, afterSpecRun, runUrl })
 
 module.exports = {
   createRun

--- a/packages/server/lib/modes/run.coffee
+++ b/packages/server/lib/modes/run.coffee
@@ -1,12 +1,15 @@
 _          = require("lodash")
+la         = require("lazy-ass")
 pkg        = require("@packages/root")
 uuid       = require("uuid")
 path       = require("path")
 chalk      = require("chalk")
+check      = require("check-more-types")
 human      = require("human-interval")
 debug      = require("debug")("cypress:server:run")
 Promise    = require("bluebird")
 logSymbols = require("log-symbols")
+
 recordMode = require("./record")
 errors     = require("../errors")
 Project    = require("../project")
@@ -265,6 +268,8 @@ renderSummaryTable = (runUrl) -> (results) ->
 
 iterateThroughSpecs = (options = {}) ->
   { specs, runEachSpec, parallel, beforeSpecRun, afterSpecRun, config } = options
+
+  la(check.maybe.fn(beforeSpecRun), "beforeSpecRun should be a function, was", beforeSpecRun)
 
   serial = ->
     Promise.mapSeries(specs, runEachSpec)
@@ -950,6 +955,9 @@ module.exports = {
           errors.throw('NO_SPECS_FOUND', config.integrationFolder, specPattern)
 
         runAllSpecs = (beforeSpecRun, afterSpecRun, runUrl) =>
+          la(check.maybe.fn(beforeSpecRun),
+            "beforeSpecRun should be a function, not", beforeSpecRun)
+
           @runSpecs({
             beforeSpecRun
             afterSpecRun

--- a/packages/server/lib/modes/run.coffee
+++ b/packages/server/lib/modes/run.coffee
@@ -1,10 +1,8 @@
 _          = require("lodash")
-la         = require("lazy-ass")
 pkg        = require("@packages/root")
 uuid       = require("uuid")
 path       = require("path")
 chalk      = require("chalk")
-check      = require("check-more-types")
 human      = require("human-interval")
 debug      = require("debug")("cypress:server:run")
 Promise    = require("bluebird")
@@ -268,8 +266,6 @@ renderSummaryTable = (runUrl) -> (results) ->
 
 iterateThroughSpecs = (options = {}) ->
   { specs, runEachSpec, parallel, beforeSpecRun, afterSpecRun, config } = options
-
-  la(check.maybe.fn(beforeSpecRun), "beforeSpecRun should be a function, was", beforeSpecRun)
 
   serial = ->
     Promise.mapSeries(specs, runEachSpec)
@@ -954,17 +950,14 @@ module.exports = {
         if not specs.length
           errors.throw('NO_SPECS_FOUND', config.integrationFolder, specPattern)
 
-        runAllSpecs = (beforeSpecRun, afterSpecRun, runUrl) =>
-          la(check.maybe.fn(beforeSpecRun),
-            "beforeSpecRun should be a function, not", beforeSpecRun)
-
+        runAllSpecs = ({ beforeSpecRun, afterSpecRun, runUrl }, parallelOverride = parallel) =>
           @runSpecs({
             beforeSpecRun
             afterSpecRun
             projectRoot
             specPattern
             socketId
-            parallel
+            parallel: parallelOverride
             browser
             project
             runUrl
@@ -1000,7 +993,8 @@ module.exports = {
             runAllSpecs
           })
         else
-          runAllSpecs()
+          ## not recording, can't be parallel
+          runAllSpecs({}, false)
 
   run: (options) ->
     electronApp

--- a/packages/server/test/e2e/7_record_spec.coffee
+++ b/packages/server/test/e2e/7_record_spec.coffee
@@ -505,6 +505,24 @@ describe "e2e record", ->
       .then ->
         expect(getRequestUrls()).to.be.empty
 
+    it "warns but does not exit when is forked pr and parallel", ->
+      process.env.CIRCLECI = "1"
+      process.env.CIRCLE_WORKFLOW_ID = "123"
+      process.env.CIRCLE_PR_NUMBER = "123"
+      process.env.CIRCLE_PR_USERNAME = "brian-mann"
+      process.env.CIRCLE_PR_REPONAME = "cypress"
+      process.env.CYPRESS_INTERNAL_E2E_TESTS = "0"
+
+      e2e.exec(@, {
+        spec: "record_pass*"
+        record: true
+        parallel: true
+        snapshot: true
+        expectedExitCode: 0
+      })
+      .then ->
+        expect(getRequestUrls()).to.be.empty
+
   context "video recording", ->
     setup(defaultRoutes, {
       video: false

